### ISSUE TITLE
setStyleUrl with callback

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#8262]
Style loading happens asynchronous and users are required to wait until the style has finished loading before altering any of the style properties. Capturing from https://github.com/mapbox/mapbox-gl-native/issues/8144 that this isn't always obvious. For example they can write code as:

```java
    mapView = (MapView) findViewById(R.id.mapView);
    mapView.onCreate(savedInstanceState);
    mapView.setStyleUrl(Style.SATELLITE);
    mapView.getMapAsync(new OnMapReadyCallback() {
      @Override
      public void onMapReady(@NonNull final MapboxMap map) {
        mapboxMap = map;
        mapboxMap.setStyleUrl(Style.MAPBOX_STREETS);
        try {
          mapboxMap.addSource(new GeoJsonSource("bus_stop",
            new URL("https://raw.githubusercontent.com/cheeaun/busrouter-sg/master/data/2/bus-stops.geojson")));
        } catch (MalformedURLException malformedUrlException) {
          Timber.e("That's not an url... ", malformedUrlException);
        }

        layer = new CircleLayer("stops_layer", "bus_stop");
        layer.setProperties(
          circleColor(Color.parseColor("#FF9800")),
          circleRadius(1.0f)
        );

        mapboxMap.addLayer(layer);
      }
 }
``` 
Executing this code would result in adding the GeoJsonSource and the CircleLayer to `Style.SATELLITE` instead of `Style.MAPBOX_STREETS`. Having `setStyleUrl(String, Callback)`would make it more obvious from API perspective + will remove the burden of writing boiler plate map change event code as:

```java
  mapView = (MapView) findViewById(R.id.mapView);
    mapView.onCreate(savedInstanceState);
    mapView.getMapAsync(new OnMapReadyCallback() {
      @Override
      public void onMapReady(@NonNull final MapboxMap map) {
        mapboxMap = map;
        mapView.addOnMapChangedListener(new MapView.OnMapChangedListener() {
          @Override
          public void onMapChanged(@MapView.MapChange int change) {
            if (change == MapView.DID_FINISH_LOADING_STYLE) {
              try {
                mapboxMap.addSource(new GeoJsonSource("bus_stop",
                  new URL("https://raw.githubusercontent.com/cheeaun/busrouter-sg/master/data/2/bus-stops.geojson")));
              } catch (MalformedURLException malformedUrlException) {
                Timber.e("That's not an url... ", malformedUrlException);
              }

              layer = new CircleLayer("stops_layer", "bus_stop");
              layer.setProperties(
                circleColor(Color.parseColor("#FF9800")),
                circleRadius(1.0f)
              );

              mapboxMap.addLayer(layer);
            }
          }
        });

        mapboxMap.setStyleUrl(Style.SATELLITE);
      }
    });

```
 
cc @mapbox/android

**Comment by 1ec5** [comment]
The iOS/macOS equivalent is being tracked in https://github.com/mapbox/mapbox-gl-native/issues/6386#issuecomment-249033669 (in conjunction with making it so that the developer sets a _style_ rather than a _style URL_).

**Comment by tobrun** [comment]
landed with #8291
